helix-tui/src/backend/crossterm.rs

```diff
@@ -23,13 +23,42 @@ use std::{
     fmt,
     io::{self, Write},
 };
+use termini::TermInfo;
 
 fn term_program() -> Option<String> {
-    std::env::var("TERM_PROGRAM").ok()
+    // Some terminals don't set $TERM_PROGRAM
+    match std::env::var("TERM_PROGRAM") {
+        Err(_) => std::env::var("TERM").ok(),
+        Ok(term_program) => Some(term_program),
+    }
 }
 fn vte_version() -> Option<usize> {
     std::env::var("VTE_VERSION").ok()?.parse().ok()
 }
+fn reset_cursor_approach(terminfo: TermInfo, fallback: &str) -> String {
+    let mut reset_str: String = Default::default();
+    if let termini::Value::Utf8String(reset_str_se) = terminfo
+        .extended_cap("Se")
+        .unwrap_or(termini::Value::Utf8String(fallback))
+    {
+        reset_str.push_str(reset_str_se);
+    }
+
+    reset_str.push_str(
+        terminfo
+            .utf8_string_cap(termini::StringCapability::CursorNormal)
+            .unwrap(),
+    );
+
+    match term_program().as_deref() {
+        // some terminals don't reset the cursor with either cnorm or Se
+        Some("tmux") => {
```

Comment:
So tmux sets both cnorm and SE and both don't contain that reset sequence? 

I guess it's because it passes that escape sequence trough but doesn't know about it itself?

I would really like to avoid special cases like this. I think what would work is if you always prepend the fallback sequence. So basically `format!("{fallback}{Se}{cnorm}")` (with this approach Se and cnorm won't need a fallback). 

It's a bit of a hack but it should work. Emitting the escape sequence multiple times is fine.

Reply:
so that approach works! although I'm confused as to why, because `printf '\033[0 q' && tput Se && tput cnorm` doesn't work... maybe the way I've implemented it in rust is wrong?